src/main_loop.rs

```diff
@@ -2374,15 +2422,74 @@ mod test {
                 .global_state_lock
                 .set_cli(mocked_cli)
                 .await;
+            let cli_args = main_loop_handler.global_state_lock.cli();
             main_loop_handler
-                .discover_peers(&mut MutableMainLoopState::new(task_join_handles))
+                .discover_peers(&mut MutableMainLoopState::new(cli_args, task_join_handles))
                 .await
                 .unwrap();
 
             let peer_discovery_sent_messages_on_peer_channel = main_to_peer_rx.try_recv().is_ok();
             assert!(peer_discovery_sent_messages_on_peer_channel);
             assert!(logs_contain("Performing peer discovery"));
         }
+
+        #[tokio::test]
+        #[traced_test]
+        async fn poorly_connected_node_tries_to_connect_to_bootstrap_node() {
+            let mut global_state_lock = setup(1, 1).await.main_loop_handler.global_state_lock;
+            let cli_args = cli_args::Args {
+                max_num_peers: 100,
+                ..Default::default()
+            };
+            global_state_lock.set_cli(cli_args).await;
+            let (peer_candidate, peer_state) =
+                set_up_peer_candidate_test(global_state_lock.clone()).await;
+
+            let net_state = &global_state_lock.lock_guard().await.net;
+            let chosen_candidate = peer_state.peer_candidate(net_state);
+            assert_eq!(Some(peer_candidate), chosen_candidate);
+        }
+
+        #[tokio::test]
+        #[traced_test]
+        async fn well_connected_node_does_not_try_to_connect_to_bootstrap_nodes() {
+            let mut global_state_lock = setup(20, 20).await.main_loop_handler.global_state_lock;
+            let cli_args = cli_args::Args {
+                max_num_peers: 42,
+                ..Default::default()
+            };
+            global_state_lock.set_cli(cli_args).await;
+            let (_, peer_state) = set_up_peer_candidate_test(global_state_lock.clone()).await;
+
+            let net_state = &global_state_lock.lock_guard().await.net;
+            let chosen_candidate = peer_state.peer_candidate(net_state);
+            assert!(chosen_candidate.is_none());
+        }
+
+        /// Create a [PotentialPeersState] and [add](PotentialPeersState::add)
+        /// one [peer candidate](PeerCandidate). The added peer candidate as
+        /// well as the resulting [PotentialPeersState] are returned.
+        ///
+        ///   * acquires `global_state_lock` for write
```

Comment:
In general, this comment is not needed for test helpers, I think. As we don't need to reason so carefully about deadlocks there.

Reply:
In my opinion, a deadlocking test is also pretty annoying. If a simple comment helps to figure out the problem, that line is well spent.

Reply:
More generally, all code is code we need to maintain. This includes test code. The standards for writing test code should be at least as high as for writing non-test code.